tutorials/haskell/snap-file-uploads/tutorial.md

```diff
@@ -0,0 +1,313 @@
+[Snap](http://snapframework.com/) framework. File uploads in Snap (that
+is,
+[`Snap.Util.FileUploads`](https://hackage.haskell.org/package/snap-core-1.0.0.0/docs/Snap-Util-FileUploads.html))
+have good documentation, but we thought it would be useful to write a
+tutorial about how to implement a basic file uploader. The idea is to
+write an application with an HTML form to upload multiple Haskell files
+(`.hs` or `.lhs`) that simply displays the contents of the uploaded
+files.
+
+This blog post was generated from a literate Haskell file that you can
+find in the
+[turbulent-sniffle](https://github.com/stackbuilders/turbulent-sniffle)
+repository. All the code was tested with [LTS Haskell
+6.10](https://www.stackage.org/lts-6.10), that is, GHC 7.10.3,
+[snap-core
+1.0.0.0](https://hackage.haskell.org/package/snap-core-1.0.0.0), and
+[snap-server
+1.0.0.0](https://hackage.haskell.org/package/snap-server-1.0.0.0).
+
+The first thing we'll do is create an `index.html` file and add an HTML
+form to upload multiple files:
+
+```html
+<form action="/" method="post" enctype="multipart/form-data">
+  <input type="file" name="files" multiple>
+  <button type="submit">Submit</button>
+</form>
+```
+
+This is just an HTML form with a file input that allows multiple files.
+We could make the input required, but we'll use Haskell to handle
+everything.
+
+In order to see exactly what we need for adding file uploads to an
+existing Haskell application, let's create a Cabal file. In this case,
+we'll call it `turbulent-sniffle.cabal`:
+
+    name: turbulent-sniffle
+    version: 0.1.0
+    build-type: Simple
+    cabal-version: >= 1.22
+
+We can now add the `index.html` file as a data file:
+
+    data-files: index.html
+
+And create an executable component:
+
+    executable turbulent-sniffle
+      main-is: Main.lhs
+      other-modules: Paths_turbulent_sniffle
+      ghc-options: -Wall
+      build-depends:
+          base >= 4.8 && < 4.9
+        , bytestring
+        , directory
+        , snap-core >= 1.0 && < 1.1
+        , snap-server >= 1.0 && < 1.1
+        , text >= 1.2 && < 1.3
+        , transformers
+      default-language: Haskell2010
+
+If you cloned the
+[turbulent-sniffle](https://github.com/stackbuilders/turbulent-sniffle)
+repository, you can now build and run the application:
+
+    $ stack build --exec turbulent-sniffle
+
+Using the following `stack.yaml` file:
+
+    extra-deps:
+    - io-streams-haproxy-1.0.0.0
+    - snap-core-1.0.0.0
+    - snap-server-1.0.0.0
+    resolver: lts-6.10
+
+Go to <http://localhost:8000/> and try to upload a Haskell file such as
+`Main.lhs`.
+
+> To myself I sadly said,  
+> ‘She's neuralgia in her head!  
+> That dear head has nothing on it!  
+> Ought she not to wear a bonnet?’
+
+Next, let's take a look at the Haskell code in `Main.lhs`. In this case,
+we only need to export the `main` function, which is probably not the
+case for a real application:
```

Comment:
In any real application `main` module exports just `main`, I'm not sure this paragraph is necessary.